### PR TITLE
support custom transformers for ts 2.3

### PR DESCRIPTION
**Pull request by longlho**
TS 2.3.1 support custom transformer so I think ts-loader should support that too

### CODE REVIEW

**Comment by johnnyreilly** [comment]
Sorry for the delay in responding - I've been pretty occupied with other things.  Would you mind giving a few more details about this please?  I'm guessing that your change is related to: https://github.com/Microsoft/TypeScript/issues/14419 ?

Having had a quick read I'm not totally clear as to the purpose of custom transformers - could you elaborate a little?

Also, it looks like you might need to resupply the PR as the changes for happypack compatibility collide with your changes.  Finally, please could also provide a test?  Thanks!

**Comment by johnnyreilly** [comment]
Ping @longlho 😄

**Comment by longlho** [comment]
@johnnyreilly sry this slipped thru my attention. I believe the purpose of custom transformers is like babel-plugin: You have more control over what's being transformed as TSC walks thru the AST. 1 example is loading external asset (https://github.com/longlho/ts-transform-img).

This allows you to `import * as img from 'foo.png'` and output would be `const img = 'base64_encoded_img'` inline. Right now I have to pipe `ts-loader` into `babel-loader` and would love to get rid of `babel-loader`.

I'll rebase the PR :)

**Comment by Igorbek** [comment]
Hey @johnnyreilly @longlho here's an example of a TS transformer for **styled-components**: https://github.com/Igorbek/typescript-plugin-styled-components. Since **ts-loader** doesn't support them, I have only section for integrating with **awesome-typescript-loader**.

**Comment by johnnyreilly** [comment]
Thanks for sharing @Igorbek - I'm hoping @longlho is going come back and finish the PR soon. He said he would so I have high hopes!

**Comment by longlho** [comment]
yeah sorry I've been tied up. Will try to wrap this up this week

**Comment by johnnyreilly** [comment]
Great - thanks @longlho!  Could we use the same API as [awesome-typescript-loader](https://github.com/s-panferov/awesome-typescript-loader/blob/a15de0a0df057bcbb248700492011539113fa139/src/interfaces.ts#L29) please?

```ts
getCustomTransformers?(): ts.CustomTransformers
```

We've a vague plan that atl and ts-loader may come together at some point.  Either way, we want it to be easy for people to move from one loader to the other as required.

**Comment by longlho** [comment]
@johnnyreilly I fixed the PR. I'm trying to find a similar test to mimic but couldn't. Can you point me to something?

**Comment by johnnyreilly** [comment]
Since there's no way for this to be tested at runtime (I think?) then it makes logical sense to create a comparison test rather than an execution test.  I'd be tempted to use this one as a basis:

https://github.com/TypeStrong/ts-loader/tree/master/test/comparison-tests/basic

Remember you only need generate the 2.3 expected output as we only run comparison tests for the latest released build.  

Do set me straight if you think this can be covered by an execution test but since it's about specifically using transformed output I'm guessing that doesn't make sense.

**Comment by johnnyreilly** [comment]
Build failure BTW: 
```
src/index.ts(189,9): error TS2345: Argument of type '{ compilerOptions: CompilerOptions; transformers: CustomTransformers; reportDiagnostics: true; fi...' is not assignable to parameter of type 'TranspileOptions'.
  Object literal may only specify known properties, and 'transformers' does not exist in type 'TranspileOptions'.
src/instances.ts(106,37): error TS2339: Property 'transformers' does not exist on type 'LoaderOptions'.
```

**Comment by Igorbek** [comment]
As a test input, can be passed a simple transformer. I can write one if you want.

**Comment by longlho** [comment]
I think build might not be passing on 2.4.0-rc :-/ I tried changing the sub version and condition to support 2.4 but a bunch failed.

**Comment by johnnyreilly** [comment]
Don't worry about 2.4 - that's a possible regression in TypeScript 2.4. see here: https://github.com/Microsoft/TypeScript/issues/16609#issuecomment-309531062

**Comment by johnnyreilly** [comment]
@Igorbek that'd be very helpful - thanks!

**Comment by johnnyreilly** [comment]
(I haven't had time to fully understand custom transformers as yet)

**Comment by johnnyreilly** [comment]
Yeah - looking at the build this all seems good. The failure seems to be TypeScript@next related and I don't think that's something that need concern us as I'm expecting that to be resolved in the compiler itself.  All we need is a test and I think we're good to go! @igorbek - if you get the chance it might be good to see if @longlho's fork works with https://github.com/Igorbek/typescript-plugin-styled-components as you would hope

**Comment by Igorbek** [comment]
Here's a simple transformer that makes all string literals uppercase:
```ts
import * as ts from 'typescript';

const transformer: ts.TransformerFactory<ts.SourceFile> = (context) => {
    const visitor: ts.Visitor = (node) => {
        if (node.kind === ts.SyntaxKind.StringLiteral) {
            const text = (node as ts.StringLiteral).text;
            if (text !== text.toUpperCase()) {
                return ts.createLiteral(text.toUpperCase());
            }
        }

        return ts.visitEachChild(node, visitor, context);
    }

    return (node) => ts.visitNode(node, visitor);
};

export default transformer;
```

**Comment by johnnyreilly** [comment]
Great! How would you expect that to be plugged into ts-loader's config?

**Comment by johnnyreilly** [comment]
Also; how would the output of that be expected to look?  Am I right when I guess that this:

```ts
const greeting = 'hi';
```

would become

```ts
const greeting = 'HI';
```
?

**Comment by longlho** [comment]
yeah I think that's the output. The way custom transformer would be passed in is something like:
```
{
  loader: 'ts-loader',
  query: {
    getCustomTransformers: () => {
      before: [require('custom-transformer')]
    }                 
  }
}
```

**Comment by Igorbek** [comment]
@johnnyreilly that's right, expected output is correct
@longlho yes, `getCustomerTransformer` should return `{ before?: Array<ts.TransformerFactory<ts.SorceFile>>; after?: Array<ts.TransformerFactory<ts.SorceFile>>; }`

```ts
getCustomTransformers: () => ({ // note parens
      before: [transformer]
})
```

**Comment by johnnyreilly** [comment]
Okay - I've had a first go at a comparison test based on @Igorbek's sample.  Unfortunately it doesn't seem to work - I'm not too sure why though... BTW I built ts-loader with TypeScript@2.4.0 but I was using TypeScript 2.3.1 to generate the test data as that's what the test pack runs against.  (So post build of ts-loader and pre-test I `npm install typescript@2.3.1`)

File: test/comparison-tests/customTransformer/webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+var path = require('path')
+
+// 1. import default from the plugin module
+var createStyledComponentsTransformer = require('./uppercaseStringLiteralTransformer').default;
+
+// 2. create a transformer;
+// the factory additionally accepts an options object which described below
+var styledComponentsTransformer = createStyledComponentsTransformer();
+
```
Comment (Igorbek) [inline test/comparison-tests/customTransformer/webpack.config.js]: The sample transformer exports the transformer itself, and not a factory function that `typescript-plugin-styled-components` does. So you don't need to call the function:
```ts
var uppercaseStringLiteralTransformer = require('./uppercaseStringLiteralTransformer').default;
```

File: test/comparison-tests/customTransformer/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        extensions: ['.ts', '.js']
+    },
+    module: {
+        rules: [
+            {
+                test: /\.ts$/,
+                loader: 'ts-loader',
+                options: {
+                    getCustomTransformers: () => ({ // note parens
+                        before: [styledComponentsTransformer]
```
Comment (Igorbek) [inline test/comparison-tests/customTransformer/webpack.config.js]: `before: [uppercaseStringLiteralTransformer]`

**Comment by johnnyreilly** [comment]
Thanks @Igorbek - just made those changes.  Looks promising!

File: test/comparison-tests/customTransformer/expectedOutput-2.3/output.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+    Asset     Size  Chunks             Chunk Names
+bundle.js  2.92 kB       0  [emitted]  main
+chunk    {0} bundle.js (main) 193 bytes [entry] [rendered]
+    [0] ./.test/customTransformer/SUBMODULE/SUBMODULE.ts 78 bytes {0} [built]
```
Comment (Igorbek) [inline test/comparison-tests/customTransformer/expectedOutput-2.3/output.txt]: transformer uppercased the module import as well :) it was unintentional, but ok ))

Comment (johnnyreilly) [inline test/comparison-tests/customTransformer/expectedOutput-2.3/output.txt]: In actual fact that is the reason the test is failing on Travis right now I think. Case insensitive Windows Vs case sensitive Linux :smile: I'll change the test so it is 1 file only (no imports) tonight and then we should have a reliable test that works cross platform. Then I think we're ready to merge!

**Comment by johnnyreilly** [comment]
Okay - so we've now got a cross platform test! However, it looks like customTransfrmers aren't applied in `transpileOnly` mode. Is that intentional?

**Comment by longlho** [comment]
hmm nope those 2 configs should be independent

**Comment by johnnyreilly** [comment]
If you look at `TranspileOptions` it doesn't appear to a valid option:

```
    interface TranspileOptions {
        compilerOptions?: CompilerOptions;
        fileName?: string;
        reportDiagnostics?: boolean;
        moduleName?: string;
        renamedDependencies?: MapLike<string>;
    }
```

**Comment by johnnyreilly** [comment]
So maybe this isn't supported for transpileOnly?

**Comment by longlho** [comment]
Maybe in 2.4 they're supporting it? (https://github.com/Microsoft/TypeScript/blob/4b3e661aaa69e53b6a5992c76f2232c9f5dace10/src/services/transpile.ts#L2-L9)

**Comment by Igorbek** [comment]
yes, 2.4rc [supports it](https://github.com/Microsoft/TypeScript/blob/v2.4-rc/src/services/transpile.ts#L8).

**Comment by johnnyreilly** [comment]
Ok - so the test was generated with 2.3 so if it was added with 2.4 that makes sense. If you use it with 2.4 it'll just work?

**Comment by johnnyreilly** [comment]
I'm happy - let's merge!

**Comment by johnnyreilly** [comment]
I'm cutting ts-loader 2.2 right now - should be released within the hour.  Thanks for the contribution!

**Comment by TomMarius** [comment]
Hey guys, how do I get to the type checker (`ts.TypeChecker`)? The method `getTypeChecker` is available only in the `ts.Program` interface and I don't see it being importable from the loader nor can I find any examples that would use the type checker. I'm probably just missing something, but I've been scratching my head for a few hours now.

**Comment by johnnyreilly** [comment]
Do you mean fork-ts-checker-webpack-plugin? It's a separate npm package: https://github.com/Realytics/fork-ts-checker-webpack-plugin

**Comment by johnnyreilly** [comment]
See here: https://github.com/TypeStrong/ts-loader/tree/master/examples/fork-ts-checker

**Comment by TomMarius** [comment]
Nonono, I meant how to get to the TypeScript's internal instance of `TypeChecker` that I use in my custom transformers. Thank you for your reply though.

**Comment by TomMarius** [comment]
See this custom transformer for example: https://github.com/kimamula/ts-transformer-enumerate/blob/master/transformer.ts

He uses argument `program` of function `visitNode` to pass down the compiler's instance of the `Program` interface. He then uses the `getTypeChecker` method and then uses the TypeChecker in the transformer. I need to get either the `TypeChecker` itself or an instance of the `Program` interface down from `ts-loader`.

**Comment by johnnyreilly** [comment]
Ah I see. To be honest I've not really done anything with transformers and so can't help. Maybe @longlho or @igorbek can...

**Comment by longlho** [comment]
@TomMarius you can't get access to the `program` currently since it's out of scope. You will need to write a custom compiler to hoist that or PR'ed to TS have that passed down as part of the `TransformationContext`. I guess `ts-loader` can pass that down as a param, althou I'm not sure how future-proof that will be.

**Comment by TomMarius** [comment]
@longlho Okay, I'll probably create a PR in the following days.

**Pull request merged**